partner: Astra DB, add indexing support for Vector Store class

hemidactylus opened the pull request:
This PR adds support to the "Indexing" options when creating Astra DB collection for fine-grained control over which fields are indexed to be later used as search filters.

The (sensible) default is to index all and only the contents of the "metadata" map (plus the vectors themselves for the ANN search of course).
But one can choose to exclude certain metadata fields from being indexed (e.g. very long unique strings), or conversely specify an allowlist - or even provide a fully custom indexing prescription fed directly to the API.

In case the collection is detected on DB already, an error is raised if the requested indexing options are incompatible. But if a legacy collection (without indexing options altogether) is detected, a warning is raised and the execution can proceed.

Added unit + integration tests about this behaviour.

sl-at-ibm committed:
adding support for indexing in the astra db vector store

hemidactylus commented:
@cbornet FYI

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores/astradb.py:
```diff
@@ -284,6 +372,82 @@ def _get_embedding_dimension(self) -> int:
+        the collection. It checks if the collection is already on DB,
+        checks the indexing options match, and takes action (errors, warnings).
+
+        Return True iff the error was related to indexing, i.e. if there is
```
```suggestion
        Return True if the error was related to indexing, i.e. if there is
```

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores/astradb.py:
```diff
@@ -284,6 +372,82 @@ def _get_embedding_dimension(self) -> int:
+        the collection. It checks if the collection is already on DB,
+        checks the indexing options match, and takes action (errors, warnings).
+
+        Return True iff the error was related to indexing, i.e. if there is
+        *no* need to re-raise the originating exception outside of this call.
```
```suggestion
        Returns:
            True if the error was related to indexing, i.e. if there is
            *no* need to re-raise the originating exception outside of this call.
```

hemidactylus replied:
Thank you, done

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores/astradb.py:
```diff
@@ -284,6 +372,82 @@ def _get_embedding_dimension(self) -> int:
+                )
+            else:
+                # the discrepancies have to do with other settings than indexing
+                raise
```
Which exception is raised ?

hemidactylus replied:
Ah, good catch! It should be the exception raised in `_[a]provision_collection`. Indeed this handler function is called only within those `except` blocks, and I believe the "last exception in the current scope" ([ref](_validate_create_collection_indexing)) is maintained across function calls.

But I agree: it has become a bit obscure after the refactoring into the separate `_validate_create_collection_indexing`. I will adjust for clarity.

cbornet replied:
Then maybe pass the original exception as argument and re-raise it ?

hemidactylus replied:
Disregard that, this should be a `return False` to let the caller know _it_ has to reraise.

sl-at-ibm committed:
Indexing management improvements and a bugfix

- removed a raise from _validate_create_collection_indexing
- rename metadata-indexing init parameters
- reformat docstring for _validate_create_collection_indexing
- generic error/warning messages for _validate_create_collection_indexing
- improved check-for-no-warning logic for indexing integration tests

sl-at-ibm committed:
lint

hemidactylus commented:
@efriis the above checks seem to show that integration tests were not performed. Is it something to be expected for a code-related PR on a partner package?

sl-at-ibm committed:
Merge branch 'master' into SL-astradb-vectorstore-indexing

sl-at-ibm committed:
wip adapting vectorstore to use astradb utilities

sl-at-ibm committed:
indexing policy management lifted to astradb utils module; storage uses it as well

cbornet commented on libs/partners/astradb/langchain_astradb/utils/astradb.py:
```diff
@@ -107,9 +117,31 @@ async def _setup_db() -> None:
+                    collections = (get_coll_response["status"] or {}).get(
+                        "collections"
+                    ) or []
+                    if not _AstraDBCollectionEnvironment.validate_indexing_policy(
```
```suggestion
                    if not self.validate_indexing_policy(
```

cbornet commented on libs/partners/astradb/langchain_astradb/utils/astradb.py:
```diff
@@ -120,12 +152,138 @@ async def _setup_db() -> None:
+            except (APIRequestError, ValueError):
+                # possibly the collection is preexisting and may have legacy,
+                # or custom, indexing settings: verify
+                get_coll_response = self.astra_db.get_collections(  # type: ignore[union-attr]
```
```suggestion
                get_coll_response = self.astra_db.get_collections(
```

cbornet commented on libs/partners/astradb/langchain_astradb/utils/astradb.py:
```diff
@@ -120,12 +152,138 @@ async def _setup_db() -> None:
+                    raise
+
+    @staticmethod
+    def validate_indexing_policy(
```
Make it private with `_` prefix ?

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores.py:
```diff
@@ -334,25 +340,20 @@ def _select_relevance_score_fn(self) -> Callable[[float], float]:
-            # async_collection not None if so is async_astra_db (constr. flow)
-            await self.async_collection.delete_many({})  # type: ignore[union-attr]
+        await self.astra_env.aensure_db_setup()
+        await self.async_collection.delete_many({})  # type: ignore[union-attr]
```
```suggestion
        await self.async_collection.delete_many({})
```

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores.py:
```diff
@@ -334,25 +340,20 @@ def _select_relevance_score_fn(self) -> Callable[[float], float]:
-        self._ensure_astra_db_client()
+        self.astra_env.ensure_db_setup()
         # self.collection is not None (by _ensure_astra_db_client)
         deletion_response = self.collection.delete_one(document_id)  # type: ignore[union-attr]
```
```suggestion
        deletion_response = self.collection.delete_one(document_id)
```

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores.py:
```diff
@@ -671,72 +666,60 @@ async def aadd_texts(
+
+        async def _handle_batch(document_batch: List[DocDict]) -> List[str]:
+            # self.async_collection is not None here for sure
+            im_result = await self.async_collection.insert_many(  # type: ignore[union-attr]
```
```suggestion
            im_result = await self.async_collection.insert_many(

cbornet commented on libs/partners/astradb/tests/integration_tests/test_vectorstores.py:
```diff
@@ -16,9 +16,12 @@
 from typing import Iterable, List, Optional, TypedDict
 
 import pytest
+from astrapy.db import AstraDB as LibAstraDB
```
No need to alias ?

cbornet commented on libs/partners/astradb/langchain_astradb/vectorstores.py:
```diff
@@ -210,110 +275,51 @@ def __init__(
-            dimension=self._get_embedding_dimension(),
-            collection_name=self.collection_name,
-            metric=self.metric,
+        self.embedding_dimension = len(
```
Do only if `self.embedding_dimension` is None ?

sl-at-ibm committed:
minor refactoring (indexing validations) after Christophe's suggestions

hemidactylus commented on libs/partners/astradb/langchain_astradb/utils/astradb.py:
```diff
@@ -107,9 +117,31 @@ async def _setup_db() -> None:
-                    collection_name, dimension=dimension, metric=metric
-                )
+
+                try:
```
I would like to defer this one. The way I understand it would have a few typechecks in the shared method, might be unwieldy (?)

sl-at-ibm committed:
Merge branch 'master' and bring the best docstrings in

cbornet reviewed (approved):
LGTM!

sl-at-ibm committed:
Merge branch 'master' into SL-astradb-vectorstore-indexing

sl-at-ibm committed:
Merge branch 'master' into SL-astradb-vectorstore-indexing

efriis commented:
Package has moved! Could you reopen here? https://github.com/langchain-ai/langchain-datastax

hemidactylus commented:
Reopened [in the new repo](https://github.com/langchain-ai/langchain-datastax/pull/6) - thank you!
